src/main/java/com/google/sps/servlets/utils/YouTubeCommentRetriever.java

```
@@ -0,0 +1,100 @@
+ */
+public class YouTubeCommentRetriever {
+  private static final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+  // Limit of comments that can be obtained in one request
+  private static final Long COMMENT_LIMIT = 100L;
+  // Parameters required by YouTube API to retrieve the comment threads
+  private static final String SNIPPET_PARAMETERS = "snippet,replies";
+  private static final String ORDER_PARAMETER = "relevance";
+
+  private static final String APPLICATION_NAME = "SAY";
+  // TODO: have dev key come from centralized location, rather than being hard-coded.
+  private static final String DEVELOPER_KEY = "AIzaSyDYfjWcy1hEe0V7AyaYzgIQm_rT-9XbiGs";
+
+  public static List<CommentThread> retrieveComments(String url, int maxComments) {
+    String nextPageToken = "";
+    int currentOverallComments = 0;
+    long commentQueryLimit = 0;
+    ArrayList<CommentThread> allComments = new ArrayList<CommentThread>();
+    try {
+      do {
+        commentQueryLimit = Math.min(COMMENT_LIMIT, maxComments - currentOverallComments);
+        YouTube.CommentThreads.List commentRequest = generateYouTubeRequest(url, commentQueryLimit);
+        if (nextPageToken != null && nextPageToken != "") {
+          commentRequest.setPageToken(nextPageToken);
+        }
+        CommentThreadListResponse commentResponse = commentRequest.execute();
+        nextPageToken = commentResponse.getNextPageToken();
+        currentOverallComments += COMMENT_LIMIT;
+        for (CommentThread thread : commentResponse.getItems()) {
+          // Extract relevant comment from CommentThreadListResponse
+          allComments.add(thread);
+        }
+        // Todo: consolidate comments into large list
+      } while (nextPageToken != null
+          && nextPageToken != ""
+          && currentOverallComments < maxComments);
+    } catch (Exception e) {
+      // In case of error, simply return what is in allComments, even if empty
```

I think we should just throw the error to YoutubeServlet to uncover the possible exceptions.

---

I personally feel that it's better to have the empty list be returned, and then the frontend can display an error message if it sees an empty list. We should probably talk about this in detail.

---

It would be nice to understand why the empty list is returned.